src/Features/LanguageServer/Protocol/LanguageServerTarget.cs

```diff
@@ -76,6 +76,52 @@ internal class LanguageServerTarget : ILanguageServerTarget
                 supportedLanguages,
                 serverKind);
             Queue.RequestServerShutdown += RequestExecutionQueue_Errored;
+
+            foreach (var metadata in RequestDispatcher.GetRegisteredMethods())
+            {
+                // Instead of concretely defining methods for each LSP method, we instead dynamically construct
+                // the generic method info from the exported handler types.  This allows us to define multiple handlers for the same method
+                // but different type parameters.  This is a key functionality to support TS external access as we do not want to couple
+                // our LSP protocol version dll to theirs.
+                //
+                // We also do not use the StreamJsonRpc support for JToken as the rpc method parameters because we want
+                // StreamJsonRpc to do the deserialization to handle streaming requests using IProgress<T>.
+                var delegatingEntryPoint = new DelegatingEntryPoint(metadata.MethodName, this);
+
+                var entryPointMethod = delegatingEntryPoint.GetType().GetMethod(nameof(DelegatingEntryPoint.EntryPointAsync));
+                entryPointMethod = entryPointMethod!.MakeGenericMethod(metadata.RequestType, metadata.ResponseType);
+
+                JsonRpc.AddLocalRpcMethod(entryPointMethod, delegatingEntryPoint, new JsonRpcMethodAttribute(metadata.MethodName) { UseSingleObjectParameterDeserialization = true });
+            }
+        }
+
+        /// <summary>
+        /// Wrapper class to hold the method and properties from the <see cref="LanguageServerTarget"/>
+        /// that the method info passed to streamjsonrpc is created from.
+        /// </summary>
+        private class DelegatingEntryPoint
```

Comment:
todo: do we really need this, or can we jsut pass around a tuple of these values...

Reply:
this type is mainly to hold onto the method name associated with the streamjsonrpc entrypoint.  Since streamjsonrpc doesn't pass the method name to us when they call our entry point.

I don't think a tuple would work, as the target needs to have both the actual entry point method as well as the method name.